File: Algorithms/CPP/423 Reconstruct Original Digits from English.cpp
class Solution {
public:
    string originalDigits(string s) {
        string result;
        array<pair<char, int>, 10> abbreviation_dict = {
            make_pair('z', 0), make_pair('w', 2), make_pair('u', 4), make_pair('x', 6), make_pair('g', 8), make_pair('o', 1), make_pair('f', 5), make_pair('s', 7), make_pair('t', 3), make_pair('i', 9)};
        string full_dict[10] = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
        array<int, 26> digit_dict = count(s);
        int counter[10] = {0};

        for (int i = 0; i < 10; ++i){
            char c = abbreviation_dict[i].first;
            int n = abbreviation_dict[i].second;
            counter[n] = digit_dict[c - 'a'];
            if (counter[n]){
                for (int i = 0; i < full_dict[n].size(); ++i){
                    digit_dict[full_dict[n].at(i) - 'a'] -= counter[n];
                }
            }
        }

        for (int i = 0; i < 10; ++i){
            result += string(counter[i], i + '0');
        }

        return result;
    }

private:
    array<int, 26> count(string s){
        array<int, 26> result = {0};
        for (int i = 0; i < s.size(); ++i){
            ++result[s[i] - 'a'];
        }

        return result;
    }
};

File: Algorithms/Python/423 Reconstruct Original Digits from English.py
import collections

class Solution:
    def originalDigits(self, s: str) -> str:
        abbreviation_dict = {'z': 0, 'w': 2, 'u': 4, 'x': 6, 'g': 8, 'o': 1, 'f': 5, 's': 7, 't': 3, 'i': 9}
        full_dict = {0: 'zero', 1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five', 6: 'six', 7: 'seven', 8: 'eight', 9: 'nine'}
        digit_dict = collections.Counter(s)
        counter = {key: 0 for key in range(10)}
        
        for key, value in abbreviation_dict.items():
            counter[value] = digit_dict[key]
            if counter[value]:
                for c in full_dict[value]:
                    digit_dict[c] = digit_dict[c] - counter[value]

        result = ''
        for key, value in counter.items():
            result = result + str(key) * value

        return result

        
